src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureKeyVaultProviderTokenCredential.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider
+{
+    internal class AzureKeyVaultProviderTokenCredential : TokenCredential
+    {
+        private AuthenticationCallback Callback { get; set; }
+
+        private string Authority { get; set; }
+
+        private string Resource { get; set; }
+
+        internal AzureKeyVaultProviderTokenCredential(AuthenticationCallback authenticationCallback, string masterKeyPath)
+        {
+            Callback = authenticationCallback;
+            HttpClient httpClient = new HttpClient();
+            HttpResponseMessage response = httpClient.GetAsync(masterKeyPath).GetAwaiter().GetResult();
+            string challenge = response?.Headers.WwwAuthenticate.FirstOrDefault()?.ToString();
+            string trimmedChallenge = ValidateChallenge(challenge);
+            string[] pairs = trimmedChallenge.Split(new string[] { "," }, StringSplitOptions.RemoveEmptyEntries);
+
+            Authority = pairs[0].Split('=')[1].Trim().Trim(new char[] { '\"' });
+            Resource = pairs[1].Split('=')[1].Trim().Trim(new char[] { '\"' });
+        }
+
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            string token = Callback.Invoke(Authority, Resource, string.Empty).GetAwaiter().GetResult();
+            return new AccessToken(token, DateTimeOffset.Now.AddHours(1));
```

Comment:
I wanted to discuss about this actually, good you brought it up.
Do we know how AKV consumes these tokens and what would be expected lifetime of this token?
Also should we cache tokens here in our implementation?

Reply:
Also, we have "ColumnEncryptionKeyCacheTtl" connection property, that sets expiration time on cache of encryption key entries. Just some thoughts to run by..

Reply:
Good questions. I do not know how AKV consumes these tokens under the hood. It may take some digging into their open source code to fully understand. As for the caching, I will need to ask around. I was not aware of that connection property before you mentioned it.

Reply:
@heaths @Petermarcu 

We'd require some assistance from Azure key Vault side to know if there's a better way to handle Expiration Time? Could you explain us how AKV (old design) set Expiration Time on Access tokens requested by calling "[AuthenticationContext.AcquireTokenAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.clients.activedirectory.authenticationcontext.acquiretokenasync)" so we can keep the experience same for Authentication Callback support in the new design?

Reply:
Key Vault uses challenge-based authentication using the initial tokens, which we cache and invalidate after they expire based on the challenge returned. See https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs for details. I'll take a look through this PR as well for more context. In general, if you use the new Azure.Security.KeyVault.* packages, you shouldn't need to implement anything on your own. The Azure.Identity package - notably `DefaultAzureCredential` - efficiently and effectively supports both developer and production workflows.

Reply:
Hi @heaths 

We are implementing our own Credential to support old "AuthenticationCallback" in our AKVProvider constructors as otherwise accepting "TokenCredential" only brings breaking changes to customers. We have added new methods to also accept TokenCredential that aligns with Azure.Security.KeyVault APIs.

This class "AzureKeyVaultProviderTokenCredential.cs" is our implementation of TokenCredential which only receives AccessToken from the user through authentication callback, and i**t's expiration time is not known to us**.

What we want to know is what would be expiration time of access tokens generated by calling "[AuthenticationContext.AcquireTokenAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.clients.activedirectory.authenticationcontext.acquiretokenasync)" APIs, is that configurable?

And what is the risk of setting expiration time to 1 hour or any other static value in our implementation above?
What would you recommend?

Reply:
The server sends those back. If the expiresOn attribute is configured by the user, you should probably leave it. In the sample I set, we merely expire/invalidate cached tokens 2 minutes before expiration, to allow for low-bandwidth scenarios.

/cc @schaabs

Reply:
Hi @heaths 

As I posted [here](https://github.com/dotnet/SqlClient/pull/630#discussion_r452541642), we don't have expirationTime from Callback method, it returns us only string "access token" that could be directly used by KeyVaultClient in old design. I did go through your design and looked at more examples but no where I found anything like we are doing here.

ChallengeBased Auth Policy receives "AccessToken" instance [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L107-L109), whereas we only get "string" access token [here](https://github.com/dotnet/SqlClient/pull/630/files/6d7eb45ffd7da077370d119128a3961c38a17916#diff-b2c8cd0d3bbeed67854fca5124d813d9R66) that is the response of old KeyVaultClient.AuthenticationCallback.

We have to construct "AccessToken" instance using only "string" access token for which expiry may be different than what we are trying to set.